src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs

```diff
@@ -1646,6 +1647,177 @@ public void GetEnumerator_UseExplicitInterfaceImplementation_EnumeratorReturnsNo
+
+            Assert.True(enumerator.MoveNext());
+            Assert.Equal(customHeaderName, enumerator.Current.Key);
+            Assert.Equal(1, enumerator.Current.Value.Count());
+            Assert.Equal("custom0", enumerator.Current.Value.ElementAt(0));
+
+            // Starting using the non-validating enumerator will NOT trigger parsing of raw values.
+            Assert.Equal(1, headers.Parser.TryParseValueCallCount);
+
+            Assert.True(enumerator.MoveNext());
+            Assert.Equal(headers.Descriptor.Name, enumerator.Current.Key);
+            Assert.Equal(2, enumerator.Current.Value.Count());
+            Assert.Equal(rawPrefix + "1", enumerator.Current.Value.ElementAt(0));
+            Assert.Equal(rawPrefix + "2", enumerator.Current.Value.ElementAt(1));
+
+            Assert.Equal(1, headers.Parser.TryParseValueCallCount);
+
+            Assert.False(enumerator.MoveNext(), "Only 2 values expected, but enumerator returns a third one.");
+        }
+
+        [Fact]
+        public void NonValidated_FirstCustomHeaderWithEmptyValueSecondKnownHeaderWithEmptyValue_EnumeratorReturnsTwoHeaders()
+        {
+            MockHeaders headers = new MockHeaders();
+            headers.Add(customHeaderName, string.Empty);
+            headers.Add(headers.Descriptor, string.Empty);
+
+            var enumerator = headers.NonValidated.GetEnumerator();
+
+            Assert.True(enumerator.MoveNext());
+            Assert.Equal(customHeaderName, enumerator.Current.Key);
+            Assert.Equal(1, enumerator.Current.Value.Count());
+            Assert.Equal(string.Empty, enumerator.Current.Value.ElementAt(0));
+
+            Assert.False(enumerator.MoveNext(), "Only the (empty) custom value should be returned.");
+        }
+
+        [Fact]
+        public void NonValidated_UseExplicitInterfaceImplementation_EnumeratorReturnsNoOfHeaders()
+        {
+            MockHeaders headers = new MockHeaders();
+            headers.Add("custom1", "customValue1");
+            headers.Add("custom2", "customValue2");
+            headers.Add("custom3", "customValue3");
+            headers.Add("custom4", "customValue4");
+
+            IEnumerable headersAsIEnumerable = headers.NonValidated;
+
+            var enumerator = headersAsIEnumerable.GetEnumerator();
+
+            KeyValuePair<string, HeaderStringValues> currentValue;
+
+            for (int i = 1; i <= 4; i++)
+            {
+                Assert.True(enumerator.MoveNext());
+                currentValue = (KeyValuePair<string, HeaderStringValues>)enumerator.Current;
+                Assert.Equal("custom" + i, currentValue.Key);
+                Assert.Equal(1, currentValue.Value.Count());
+            }
+
+            Assert.False(enumerator.MoveNext(), "Only 4 values expected, but enumerator returns a fifth one.");
+        }
+
+        [Fact]
+        public void NonValidated_AddAndUpdateSingleValueHeader_EnumeratorReturnsLatestValue()
```

Comment:
Can we add tests similar to this one and the next two, except that we start with a raw value instead of parsed values?

Reply:
Done. Added tests
- [NonValidated_AddAndUpdateRawSingleValueHeader_EnumeratorReturnsLatestValue](https://github.com/dotnet/runtime/blob/9c860a47d50cedbbe60f49ef891cd0b453bb0e9e/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs#L1823)
- [NonValidated_AddAndRemoveRawSingleValueHeader_EnumeratorReturnsNoValue](https://github.com/dotnet/runtime/blob/9c860a47d50cedbbe60f49ef891cd0b453bb0e9e/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs#L1848)
- [NonValidated_AddAndRemoveRawValueToMultivalueHeader_EnumeratorReturnsLatestParsedValue](https://github.com/dotnet/runtime/blob/9c860a47d50cedbbe60f49ef891cd0b453bb0e9e/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs#L1880)